waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,512 @@
+  )
+
+  while true:
+    let response = (await wakuArchive.findMessages(query)).valueOr:
+      return err($error)
+
+    for i in 0 ..< response.hashes.len:
+      let hash = response.hashes[i]
+      let topic = response.topics[i]
+      let msg = response.messages[i]
+
+      self.messageIngress(topic, hash, msg)
+
+    if response.cursor.isNone():
+      break
+
+    query.cursor = response.cursor
+
+  return ok()
+
+proc new*(
+    T: type WakuSync,
+    peerManager: PeerManager,
+    maxFrameSize: int = DefaultMaxFrameSize,
+    syncRange: timer.Duration = DefaultSyncRange,
+    syncInterval: timer.Duration = DefaultSyncInterval,
+    relayJitter: Duration = DefaultGossipSubJitter,
+    wakuArchive: WakuArchive,
+    wakuStoreClient: WakuStoreClient,
+    pruneCallback: Option[PruneCallback] = none(PruneCallback),
+    transferCallback: Option[TransferCallback] = none(TransferCallback),
+): Future[Result[T, string]] {.async.} =
+  let storage = NegentropyStorage.new().valueOr:
+    return err("negentropy storage creation failed")
+
+  var sync = WakuSync(
+    storage: storage,
+    peerManager: peerManager,
+    maxFrameSize: maxFrameSize,
+    syncInterval: syncInterval,
+    syncRange: syncRange,
+    relayJitter: relayJitter,
+    pruneOffset: syncInterval div 100,
+  )
+
+  sync.initProtocolHandler()
+
+  sync.pruneCallBack = pruneCallback
+
+  if sync.pruneCallBack.isNone():
+    let res = sync.createPruneCallback(wakuArchive)
+
+    if res.isErr():
+      error "pruning callback creation error", error = res.error
+    else:
+      sync.pruneCallBack = some(res.get())
+
+  sync.transferCallBack = transferCallback
+
+  if sync.transferCallBack.isNone():
+    let res = sync.createTransferCallback(wakuArchive, wakuStoreClient)
+
+    if res.isErr():
+      error "transfer callback creation error", error = res.error
+    else:
+      sync.transferCallBack = some(res.get())
+
+  let res = await sync.initFillStorage(wakuArchive)
+  if res.isErr():
+    warn "will not sync messages before this point in time", error = res.error
+
+  info "WakuSync protocol initialized"
+
+  return ok(sync)
+
+proc periodicSync(self: WakuSync, callback: TransferCallback) {.async.} =
+  debug "periodic sync initialized", interval = $self.syncInterval
+
+  # infinite loop
+  while true:
+    await sleepAsync(self.syncInterval)
+
+    debug "periodic sync started"
+
+    var
+      hashes: seq[WakuMessageHash]
+      peer: RemotePeerInfo
+      tries = 3
+
+    while true:
+      let res = (await self.storeSynchronization()).valueOr:
+        # we either try again or log an error and break
+        if tries > 0:
+          tries -= 1
+          await sleepAsync(RetryDelay)
+          continue
+        else:
+          error "sync failed", error = $error
+          break
+
+      hashes = res[0]
+      peer = res[1]
+      break
+
+    if hashes.len > 0:
+      tries = 3
+      while true:
+        (await callback(hashes, peer.peerId)).isOkOr:
```

Comment:
We don't need to pass a callback. I'd suggest having a private proc `performTransfer`.
```suggestion
        (await performTransfer(hashes, peer.peerId)).isOkOr:
```

Reply:
Only true if we don't care about mocking the callback for testing. Lets keep the callback for QA.